src/main/java/org/kohsuke/github/GitHub.java

```diff
@@ -315,31 +316,61 @@ public void setConnector(HttpConnector connector) {
         }
+
+        return this.rateLimit = rateLimit;
     }
 
-    /*package*/ void updateRateLimit(@Nonnull GHRateLimit observed) {
+    /**
+     * Update the Rate Limit with the latest info from response header.
+     * Due to multi-threading requests might complete out of order, we want to pick the one with the most recent info from the server.
+     *
+     * @param observed {@link GHRateLimit.Record} constructed from the response header information
+     */
+    void updateCoreRateLimit(@Nonnull GHRateLimit.Record observed) {
         synchronized (headerRateLimitLock) {
-            if (headerRateLimit == null
-                    || headerRateLimit.getResetDate().getTime() < observed.getResetDate().getTime()
-                    || headerRateLimit.remaining > observed.remaining) {
-                headerRateLimit = observed;
+            if (headerRateLimit == null || shouldReplace(observed, headerRateLimit.getCore())) {
+                headerRateLimit = GHRateLimit.fromHeaderRecord(observed);
                 LOGGER.log(FINE, "Rate limit now: {0}", headerRateLimit);
             }
         }
     }
 
+    /**
+     * Update the Rate Limit with the latest info from response header.
+     * Due to multi-threading requests might complete out of order, we want to pick the one with the most recent info from the server.
+     * Header date is only accurate to the second, so we look at the information in the record itself.
+     *
+     * {@link GHRateLimit.UnknownLimitRecord}s are always replaced by regular {@link GHRateLimit.Record}s.
+     * Regular {@link GHRateLimit.Record}s are never replaced by {@link GHRateLimit.UnknownLimitRecord}s.
+     * Candidates with resetEpochSeconds later than current record are more recent.
+     * Candidates with the same reset and a lower remaining count are more recent.
+     * Candidates with an earlier reset are older.
+     *
+     * @param candidate {@link GHRateLimit.Record} constructed from the response header information
+     * @param current the current {@link GHRateLimit.Record} record
+     */
+    static boolean shouldReplace(@Nonnull GHRateLimit.Record candidate, @Nonnull GHRateLimit.Record current) {
+        if (candidate instanceof GHRateLimit.UnknownLimitRecord && !(current instanceof GHRateLimit.UnknownLimitRecord)) {
+            // Unknown candidate never replaces a regular record
+            return false;
+        } else if (current instanceof GHRateLimit.UnknownLimitRecord && !(candidate instanceof GHRateLimit.UnknownLimitRecord)) {
+            // Any real record should replace an unknown Record.
+            return true;
+        } else {
+            // records of the same type compare to each other as normal.
+            return current.getResetEpochSeconds() < candidate.getResetEpochSeconds()
+                || (current.getResetEpochSeconds() == candidate.getResetEpochSeconds() && current.getRemaining() > candidate.getRemaining());
```

Comment:
@PauloMigAlmeida  
This is the check that your question got me to fix. 
A recorder with a later reset is for sure newer that one with an earlier one. 
Only if they have the same reset should we take the one with the lower remaining count.  
Previous implementation would take the lower remaining count candidate even if it had an earlier (older) reset date.  
Thanks!
We should always update to a newer record - one w

Reply:
It makes sense! I like the train of thought you used for solving it.